{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5cff650e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcfpy\n",
    "import pandas as pd\n",
    "import os\n",
    "import sqlite3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1268b5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = sqlite3.connect(\n",
    "    os.path.join(os.environ['VSC_SCRATCH'], 'test.sqlite'))\n",
    "vcf = '/lustre1/project/stg_00079/teaching/data/snps.annotated.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bdf2a4a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "db  : <sqlite3.Connection object at 0x2ac10a80c540>\n",
      "vcf : /lustre1/project/stg_00079/teaching/data/snps.annotated.vcf\n"
     ]
    }
   ],
   "source": [
    "print(f\"db  : {db}\")\n",
    "print(f\"vcf : {vcf}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "02d6b2c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snp records 3505\n",
      "call records 7010\n",
      "effect records 19640\n"
     ]
    }
   ],
   "source": [
    "# Create a VCF reader - this parses through the VCF file\n",
    "reader = vcfpy.Reader.from_path(vcf)\n",
    "\n",
    "# Empty lists to store the data in\n",
    "snp_records = []\n",
    "effect_records = []\n",
    "call_records = []\n",
    "\n",
    "# These are the columns from a SNPeff record - easy to later assign column names\n",
    "effect_rec_names = \"\"\"snp allele effect impact gene gene_id feature_type feature_id \n",
    "                      biotype rank hgvs.c hgvs.p cdna_pos cds_pos prot_pos distance_to_feature\n",
    "                      messages\"\"\".split()\n",
    "    \n",
    "# start parsing through the VCF:\n",
    "for i, record in enumerate(reader):\n",
    "    \n",
    "    # we used vt decompose - so I expect NO multiallelic SNPs\n",
    "    assert len(record.ALT) == 1 \n",
    "    \n",
    "    # this is the ALT allele (first and only)\n",
    "    alt = record.ALT[0]\n",
    "    \n",
    "    # compose a SNP name for joins later on\n",
    "    snp_name = f\"{record.CHROM}:{record.POS}:{record.REF}:{alt.value}\"\n",
    "    \n",
    "    # Store a SNP record - I collect everything now in a list of dictionaries\n",
    "    # that I will later convert to a Pandas dataframe\n",
    "    snp_records.append(\n",
    "        dict(snp = snp_name,\n",
    "             chrom = record.CHROM,\n",
    "             pos = record.POS,\n",
    "             quality = record.QUAL,\n",
    "             ref = record.REF,\n",
    "             type = alt.type,\n",
    "             alt = alt.value))\n",
    "\n",
    "    # Now I will parse through the individual calls\n",
    "    for call_record in record.calls:\n",
    "        \n",
    "        #name of the sample (TLE66_N or TLE66_T)\n",
    "        sample = os.path.basename(call_record.sample)\n",
    "        sample = sample.replace('.bam', '')\n",
    "\n",
    "        # simple genotype - -1 means not called, otherwise it's the number of ALT alleles observer\n",
    "        # this helps in later queries \n",
    "        # Note - this is not perfect - it does ignore uncalled loci (./.)\n",
    "        genotype_simple = call_record.data['GT'].count('1')\n",
    "        \n",
    "        call_records.append(\n",
    "            dict(snp = snp_name,\n",
    "                 sample = sample,\n",
    "                 genotype = call_record.data['GT'], \n",
    "                 genotype_simple = genotype_simple))\n",
    "\n",
    "    # and parse through all snpEff annotations\n",
    "    for ann in record.INFO['ANN']:\n",
    "        ann = ann.split('|')\n",
    "        # create a dictionary of all fields\n",
    "        eff_record = dict(zip(effect_rec_names, [snp_name] + ann))\n",
    "        # convert distance to feature to integer\n",
    "        try:\n",
    "            eff_record[intfield] = int(eff_record.get(intfield))\n",
    "        except:\n",
    "            eff_record[intfield] = -1\n",
    "            \n",
    "        effect_records.append(eff_record)\n",
    "\n",
    "    #for debugging\n",
    "    #if i > 4000:\n",
    "    #    break\n",
    "\n",
    "#convert lists of dicts to a DataFrame\n",
    "snp_records = pd.DataFrame.from_records(snp_records)\n",
    "call_records = pd.DataFrame.from_records(call_records)\n",
    "effect_records = pd.DataFrame.from_records(effect_records)\n",
    "\n",
    "#Save to disk\n",
    "print('snp records', snp_records.to_sql('snp', db, if_exists='replace', index=False))\n",
    "print('call records', call_records.to_sql('snp_call', db, if_exists='replace', index=False))\n",
    "print('effect records', effect_records.to_sql('snp_effect', db, if_exists='replace', index=False))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7a05f02b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>quality</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127578816</td>\n",
       "      <td>36.1174</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127578974</td>\n",
       "      <td>422.7380</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127579080</td>\n",
       "      <td>172.0220</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127663498:C:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127663498</td>\n",
       "      <td>66.0472</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127674824:G:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127674824</td>\n",
       "      <td>24.6981</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp chrom        pos   quality ref type alt\n",
       "0  chr9:127578816:C:T  chr9  127578816   36.1174   C  SNV   T\n",
       "1  chr9:127578974:A:G  chr9  127578974  422.7380   A  SNV   G\n",
       "2  chr9:127579080:A:G  chr9  127579080  172.0220   A  SNV   G\n",
       "3  chr9:127663498:C:T  chr9  127663498   66.0472   C  SNV   T\n",
       "4  chr9:127674824:G:T  chr9  127674824   24.6981   G  SNV   T"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"SELECT * FROM snp LIMIT 5\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "69faffbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>genotype_simple</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp   sample genotype  genotype_simple\n",
       "0  chr9:127578816:C:T  TLE66_N      0/1                1\n",
       "1  chr9:127578816:C:T  TLE66_T      0/1                1\n",
       "2  chr9:127578974:A:G  TLE66_N      0/1                1\n",
       "3  chr9:127578974:A:G  TLE66_T      0/1                1\n",
       "4  chr9:127579080:A:G  TLE66_N      0/1                1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"SELECT * FROM snp_call LIMIT 5\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d51df9f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>allele</th>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>effect</th>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>impact</th>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>STXBP1</td>\n",
       "      <td>NIBAN2</td>\n",
       "      <td>NIBAN2</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>NIBAN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>ENSG00000136854</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>ENSG00000136854</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_type</th>\n",
       "      <td>transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <td>ENST00000637521.2</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "      <td>ENST00000637521.2</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biotype</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rank</th>\n",
       "      <td></td>\n",
       "      <td>1/13</td>\n",
       "      <td>1/14</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hgvs.c</th>\n",
       "      <td>c.-724C&gt;T</td>\n",
       "      <td>c.16+106G&gt;A</td>\n",
       "      <td>c.-37T&gt;C</td>\n",
       "      <td>c.-566A&gt;G</td>\n",
       "      <td>c.-143T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hgvs.p</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cdna_pos</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cds_pos</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prot_pos</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>distance_to_feature</th>\n",
       "      <td>554</td>\n",
       "      <td>-1</td>\n",
       "      <td>37</td>\n",
       "      <td>396</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>messages</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         0                   1  \\\n",
       "snp                     chr9:127578816:C:T  chr9:127578816:C:T   \n",
       "allele                                   T                   T   \n",
       "effect               upstream_gene_variant      intron_variant   \n",
       "impact                            MODIFIER            MODIFIER   \n",
       "gene                                STXBP1              NIBAN2   \n",
       "gene_id                    ENSG00000136854     ENSG00000136830   \n",
       "feature_type                    transcript          transcript   \n",
       "feature_id               ENST00000637521.2   ENST00000373314.7   \n",
       "biotype                     protein_coding      protein_coding   \n",
       "rank                                                      1/13   \n",
       "hgvs.c                           c.-724C>T         c.16+106G>A   \n",
       "hgvs.p                                                           \n",
       "cdna_pos                                                         \n",
       "cds_pos                                                          \n",
       "prot_pos                                                         \n",
       "distance_to_feature                    554                  -1   \n",
       "messages                                                         \n",
       "\n",
       "                                       2                      3  \\\n",
       "snp                   chr9:127578974:A:G     chr9:127578974:A:G   \n",
       "allele                                 G                      G   \n",
       "effect               5_prime_UTR_variant  upstream_gene_variant   \n",
       "impact                          MODIFIER               MODIFIER   \n",
       "gene                              NIBAN2                 STXBP1   \n",
       "gene_id                  ENSG00000136830        ENSG00000136854   \n",
       "feature_type                  transcript             transcript   \n",
       "feature_id             ENST00000373314.7      ENST00000637521.2   \n",
       "biotype                   protein_coding         protein_coding   \n",
       "rank                                1/14                          \n",
       "hgvs.c                          c.-37T>C              c.-566A>G   \n",
       "hgvs.p                                                            \n",
       "cdna_pos                                                          \n",
       "cds_pos                                                           \n",
       "prot_pos                                                          \n",
       "distance_to_feature                   37                    396   \n",
       "messages                                                          \n",
       "\n",
       "                                         4  \n",
       "snp                     chr9:127579080:A:G  \n",
       "allele                                   G  \n",
       "effect               upstream_gene_variant  \n",
       "impact                            MODIFIER  \n",
       "gene                                NIBAN2  \n",
       "gene_id                    ENSG00000136830  \n",
       "feature_type                    transcript  \n",
       "feature_id               ENST00000373314.7  \n",
       "biotype                     protein_coding  \n",
       "rank                                        \n",
       "hgvs.c                           c.-143T>C  \n",
       "hgvs.p                                      \n",
       "cdna_pos                                    \n",
       "cds_pos                                     \n",
       "prot_pos                                    \n",
       "distance_to_feature                     91  \n",
       "messages                                    "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# .T transposes - easier viewing\n",
    "pd.read_sql(\"\"\"SELECT * FROM snp_effect LIMIT 5\n",
    "            \"\"\", db).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55061585",
   "metadata": {},
   "source": [
    "#### Find high impact variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "84ee65fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>gene</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:128632018:A:G</td>\n",
       "      <td>SPTAN1</td>\n",
       "      <td>splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:133257521:T:TC</td>\n",
       "      <td>ABO</td>\n",
       "      <td>splice_acceptor_variant&amp;splice_donor_variant&amp;i...</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:133257521:T:TC</td>\n",
       "      <td>ABO</td>\n",
       "      <td>splice_acceptor_variant&amp;splice_donor_variant&amp;i...</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:135077073:G:GAA</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:135077150:GCA:G</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    snp    gene  \\\n",
       "0    chr9:128632018:A:G  SPTAN1   \n",
       "1   chr9:133257521:T:TC     ABO   \n",
       "2   chr9:133257521:T:TC     ABO   \n",
       "3  chr9:135077073:G:GAA   OLFM1   \n",
       "4  chr9:135077150:GCA:G   OLFM1   \n",
       "\n",
       "                                              effect impact  \n",
       "0             splice_acceptor_variant&intron_variant   HIGH  \n",
       "1  splice_acceptor_variant&splice_donor_variant&i...   HIGH  \n",
       "2  splice_acceptor_variant&splice_donor_variant&i...   HIGH  \n",
       "3                                 frameshift_variant   HIGH  \n",
       "4                                 frameshift_variant   HIGH  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"SELECT snp.snp, snp_effect.gene, snp_effect.effect, snp_effect.impact\n",
    "                 FROM snp, snp_effect\n",
    "                WHERE snp.snp = snp_effect.snp\n",
    "                  AND snp_effect.impact == 'HIGH'\n",
    "                LIMIT 5\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "220fc068",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no SNPs found: 67\n",
      " * per impact:\n",
      "  -     2 : HIGH\n",
      "  -    65 : MODIFIER\n",
      " * per effect:\n",
      "  -    14 : downstream_gene_variant\n",
      "  -    46 : intron_variant\n",
      "  -     5 : non_coding_transcript_exon_variant\n",
      "  -     2 : splice_acceptor_variant&splice_donor_variant&intron_variant\n"
     ]
    }
   ],
   "source": [
    "def find_around_gene(gene):\n",
    "    res = pd.read_sql(f\"\"\"SELECT snp.snp, snp_effect.gene, snp_effect.effect, snp_effect.impact\n",
    "                          FROM snp, snp_effect\n",
    "                         WHERE snp.snp = snp_effect.snp\n",
    "                           AND snp_effect.gene == '{gene}'\"\"\", db)\n",
    "    print(\"no SNPs found:\", len(res))\n",
    "    print(\" * per impact:\")\n",
    "    for k, v in res.groupby('impact')['snp'].agg(len).iteritems():\n",
    "        print(f\"  - {v:5} : {k}\")\n",
    "    print(\" * per effect:\")\n",
    "    for k, v in res.groupby('effect')['snp'].agg(len).iteritems():\n",
    "        print(f\"  - {v:5} : {k}\")\n",
    "    \n",
    "find_around_gene('ABO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "14541377",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>genotype</th>\n",
       "      <th>gene</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>./.</td>\n",
       "      <td>1/1</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>./.</td>\n",
       "      <td>1/1</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp   sample   sample genotype genotype  gene  \\\n",
       "0  chr9:127802988:A:G  TLE66_N  TLE66_T      1/1      ./.  FPGS   \n",
       "1  chr9:127802988:A:G  TLE66_T  TLE66_N      ./.      1/1  FPGS   \n",
       "2  chr9:127802988:A:G  TLE66_N  TLE66_T      1/1      ./.  FPGS   \n",
       "3  chr9:127802988:A:G  TLE66_T  TLE66_N      ./.      1/1  FPGS   \n",
       "4  chr9:127802988:A:G  TLE66_N  TLE66_T      1/1      ./.  FPGS   \n",
       "\n",
       "             effect    impact  \n",
       "0  missense_variant  MODERATE  \n",
       "1  missense_variant  MODERATE  \n",
       "2  missense_variant  MODERATE  \n",
       "3  missense_variant  MODERATE  \n",
       "4  missense_variant  MODERATE  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"SELECT snp.snp, \n",
    "                      sc1.sample, sc2.sample, \n",
    "                      sc1.genotype, sc2.genotype,\n",
    "                      snp_effect.gene, snp_effect.effect, snp_effect.impact\n",
    "                 FROM snp, snp_call as sc1, snp_call as sc2, snp_effect\n",
    "                WHERE snp.snp = sc1.snp\n",
    "                  AND snp.snp = sc2.snp\n",
    "                  AND sc1.genotype != sc2.genotype\n",
    "                  AND snp.snp = snp_effect.snp\n",
    "                  AND snp_effect.impact != 'LOW'\n",
    "                  AND snp_effect.impact != 'MODIFIER'\n",
    "                LIMIT 5\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9872d17",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
